sdk/core/azure-core/src/main/java/com/azure/core/http/policy/AddHeadersPolicy.java

```diff
@@ -10,23 +10,36 @@
 import com.azure.core.http.HttpResponse;
 import reactor.core.publisher.Mono;
 
+import java.util.Objects;
+import java.util.function.Supplier;
+
 /**
  * The pipeline policy that adds a particular set of headers to HTTP requests.
  */
 public class AddHeadersPolicy implements HttpPipelinePolicy {
-    private final HttpHeaders headers;
+    private final Supplier<HttpHeaders> headersSupplier;
```

Comment:
As discussed, using suppliers may not be thread-safe here. So, consider looking into using `Context`.

Reply:
May be we should clarify this in our pipeline guidelines: An instance of a policy is shared across multiple calls, hence concurrently called by the http-client-threads sharing the same pipeline. This implicitly means, any business logic externalized from the policy, through Function, Supplier, are not thread-safe as well. Any state management should be happen through HttpPipelineCallContext object.

Reply:
Will add POC for this approach in same PR.